### PR TITLE
remove `neon_local --pageserver-config-override`

### PR DESCRIPTION
**Pull request by problame**
Preceding PR https://github.com/neondatabase/neon/pull/7613 reduced the usage of `--pageserver-config-override`.

This PR builds on top of that work and fully removes the `neon_local --pageserver-config-override`.

Tests that need a non-default `pageserver.toml` control it using two options:

1. Specify `NeonEnvBuilder.pageserver_config_override` before `NeonEnvBuilder.init_start()`. This uses a new `neon_local init --pageserver-config` flag.
2. After `init_start()`: `env.pageserver.stop()` + `NeonPageserver.edit_config_toml()` + `env.pageserver.start()`

A few test cases were using `env.pageserver.start(overrides=("--pageserver-config-override...",))`.
I changed them to use one of the options above. 

Future Work
-----------

The `neon_local init --pageserver-config` flag still uses `pageserver --config-override` under the hood. In the future, neon_local should just write the `pageserver.toml` directly.

The `NeonEnvBuilder.pageserver_config_override` field should be renamed to `pageserver_initial_config`. Let's save this churn for a separate refactor commit.

### CODE REVIEW

File: test_runner/performance/test_branch_creation.py [hunk 1]
```diff
@@ -140,10 +140,13 @@ def test_branch_creation_many(neon_compare: NeonCompare, n_branches: int, shape:
 
     # start without gc so we can time compaction with less noise; use shorter
     # period for compaction so it starts earlier
+    def patch_default_tenant_config(config):
+        config["compaction_period"] = "3s"
+        config["gc_period"] = "0s"
+        return True
+
+    env.pageserver.edit_config_toml(patch_default_tenant_config)
```
Comment (koivunej) [inline test_runner/performance/test_branch_creation.py:148]: this looks like it edits the main keys of pageserver config, not the tenant config anymore.

File: test_runner/fixtures/neon_fixtures.py [hunk 2]
```diff
@@ -2415,9 +2423,47 @@ def tenant_dir(
             return self.workdir / "tenants"
         return self.workdir / "tenants" / str(tenant_shard_id)
 
+    @property
+    def config_toml_path(self) -> Path:
+        return self.workdir / "pageserver.toml"
+
+    def edit_config_toml(self, edit_fn: Callable[[Dict[str, Any]], bool]):
```
Comment (problame) [inline test_runner/fixtures/neon_fixtures.py:2430]: Make return nothing, make throw exception if no change wanted

File: test_runner/performance/test_branch_creation.py (hunk 1 above)
Comment (problame) [reply test_runner/performance/test_branch_creation.py]: 66fc413a72685caef015d3a257b4694419d92945

File: test_runner/fixtures/neon_fixtures.py (hunk 2 above)
Comment (problame) [reply test_runner/fixtures/neon_fixtures.py]: 66fc413a72685caef015d3a257b4694419d92945

**Comment by koivunej** [comment]
Congrats! This had so large test report message that github refused it :) This is the first I've seen such.

From the logs this seems to be quite usual failure: https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7614/8967600476/index.html#suites/7745dadbd815ab87f5798aa881796f47/baabb8517aad95b1

**Review by koivunej — APPROVED** [review]

**Comment by problame** [comment]
The compat tests are failing because the old `neon_local` binary doesn't have the `--pageserver-config` flag. 

This is immaterial to

1. neon_local users unless they uses `--pageserver-config-override`
2. out production usage, because we don't use `neon_local` in production

**Comment by problame** [comment]
Extracted fix for test_forward_compatibility into PR https://github.com/neondatabase/neon/pull/7637 , this one will be rebased after it gets merged.

**Pull request merged**